File: exercise/leetcode/0036/0036.py

from typing import List


from typing import List

class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rows = [set() for _ in range(9)]
        cols = [set() for _ in range(9)]
        blocks = [set() for _ in range(9)]

        for i in range(len(board)):
            line = board[i]
            for j in range(len(line)):
                val = line[j]
                if val == '.':
                    continue
                val = int(val)

                # in row ?
                if val in rows[i]:
                    return False
                rows[i].add(val)

                # in col ?
                if val in cols[j]:
                    return False
                cols[j].add(val)

                # in block ?
                block_idx = i // 3 + 3 * (j // 3)
                if val in blocks[block_idx]:
                    return False
                blocks[block_idx].add(val)
        return True

File: exercise/leetcode/0036/0036.cc
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        int rows[9][10] = {0};
        int cols[9][10] = {0};
        int block[9][10] = {0};

        for (int i=0; i<9; ++i) {
            vector<char>& line = board[i];
            for (int j=0; j<9; ++j) {
                char v = line[j];
                if (v == '.') continue;
                int val = v - '0';

                if (rows[i][val] == 1) return false;
                rows[i][val] = 1;

                if (cols[j][val] == 1) return false;
                cols[j][val] = 1;

                int block_idx = i / 3 + 3 * (j / 3);
                if (block[block_idx][val] == 1) return false;
                block[block_idx][val] = 1;
            }
        }
        return true;
    }
};